crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -1223,13 +1238,19 @@ impl<P: PhaseItem, const I: usize> RenderCommand<P> for SetMeshBindGroup<I> {
             );
             return RenderCommandResult::Failure;
         };
+
         let mut set_bind_group = |indices: &[u32]| pass.set_bind_group(I, bind_group, indices);
-        let mesh_index = mesh_index.index();
-        match (skin_index, morph_index) {
-            (None, None) => set_bind_group(&[mesh_index]),
-            (Some(skin), None) => set_bind_group(&[mesh_index, skin.index]),
-            (None, Some(morph)) => set_bind_group(&[mesh_index, morph.index]),
-            (Some(skin), Some(morph)) => set_bind_group(&[mesh_index, skin.index, morph.index]),
+        match (batch_indices.dynamic_offset, skin_index, morph_index) {
+            (None, None, None) => set_bind_group(&[]),
+            (None, Some(skin), None) => set_bind_group(&[skin.index]),
+            (None, None, Some(morph)) => set_bind_group(&[morph.index]),
+            (None, Some(skin), Some(morph)) => set_bind_group(&[skin.index, morph.index]),
+            (Some(mesh_index), None, None) => set_bind_group(&[mesh_index]),
+            (Some(mesh_index), Some(skin), None) => set_bind_group(&[mesh_index, skin.index]),
+            (Some(mesh_index), None, Some(morph)) => set_bind_group(&[mesh_index, morph.index]),
+            (Some(mesh_index), Some(skin), Some(morph)) => {
+                set_bind_group(&[mesh_index, skin.index, morph.index]);
+            }
```

Comment:
```suggestion
        let mut bindgroup_indices: SmallVec<[u32; 3]> = SmallVec::with_capacity(3);
        if let Some(mesh_index) = batch_indices.dynamic_offset {
            bindgroup_indices.push(mesh_index);
        }
        if let Some(skin_index) = skin_index {
            bindgroup_indices.push(skin_index.index);
        }
        if let Some(morph_index) = morph_index {
            bindgroup_indices.push(morph_index.index);
        }
        pass.set_bind_group(I, bind_group, &bindgroup_indices);
```
only a little bit cleaner but at least it won't grow exponentially.

Reply:
I'd like us to double check the generated assembly and make sure this isn't less performant, considering how hot of a function this is. Should be fine though.

Reply:
I had written it like this initially but then saw the code changes on `main` and adopted it. I'm happy with and prefer the `SmallVec` approach though.

Reply:
Done.

Reply:
it is actually a lot more code in godbolt (using 1.59 since later versions can't use smallvec apparently).

the original was [165 instructions](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATKUcAZPAZMADl3ACNMYgkAVlIAB1QFQjsGFzcPb3jE5IEAoNCWCKiuWKtMGxShAiZiAjT3Tx8yioEqmoI8kPDImMtq2vqMpv6OwK7CnpKASktUE2Jkdg44kzCAaioGDYYuCCYQNYB5ONsBAFIAZicTC69L7FI1sIPj04ZL69v7x%2BQXk5SPjc7hdsFM1mcAOwAITOGgAnLCAIJrFGolH0AhrFgmTFKAgAfTCgXQ%2BOAxDmcXBFwAIuCIU5iXhFgoDmcvAA2M7RKFArnUyFONZxYiBAi0BhgSBsu7QkCXABikP5Xh8a0ZzKmlxhGmRaNRLCYBGQCDWe0eYR%2BYMh2t1er1EGCAkwj0dQRdTqtNPuazxhOJpPJJjiEDZnO5fJmSLtdodTseQlQbAgCgA1oEZmtXZhPfyQT7MASiQwSWSKSGOVyoanAhHSFHo2jY27M3G1gmkywSHEEFMc97fUWS4Hg6HK53iN3a/WGyim8624nMMm0wwM%2B2l%2BPu72qbnsPnC/7S0Hy2GqyvHpuEFOdTPURB1xA2AoEPjiZhVBms%2B6gn28wPD8OJ5jpgz6vsW77XraM73ouj4gS%2Bb4fvGsHVqu37Zju/YFn6xYBmWo7ck%2BCHgaojyoZBt6zg%2BRFgVgSEts21Fdj2v57v%2BuFHiOFaEfBtHvhezEUZRMEdrxiFrihK4SR2zHbpcu50jalF6uxQ74dxUI0YhZHnlignRNSmoXEplFKkisIIhC/I6kqHAzLQnDRLwngcFopCoJwABKZi4nMCyYOCXgXDwpAEJodkzCmIAXBcAB0MUJYlSXsvonCSM54XuZwvAshooXhTMcCwEgaAsHEdCROQlCleV9BRMQJRwnlWAAG5MpgABqeCYAA7scjCcCFNC0AQkQshAzyubwRY1AAnoN02BHNhxhNo5RhdwvClWwgiHAwtDzVNpBYGEJjAE4Yi0Cym3HZgBpGOIR34MQ614C1IGZe%2B5Q4ksIWipgDlHbQeBhMQc0uFgmUECKLALTMVAGMAChdb1/UuUNsgiGI7BSDIgiKCo6hHboXD6IYxg%2BfoIMsrAzBsCACgGrQtDvcgpDvVE3AABxaL27n/AI10ALSCqgJxC/Q720PJFy8KgHMilgNMQDMzRvA4xaDJ4pP%2BGMBRFHoCRJG8WuG9kbydPrkyWADr2tCMpuk2rlQjJb3TFH07SO57tRuxMxSq35iwSPZjkZUdHkcGspjmMgaxcLF0RwrFGimrghAkIFwVTLwG285F0VxVzFwaFzkjRBCEJcFIkheBCKWA%2BlpAuW5kc5SAeV53ZpBFYgIBzAQKwWBQEA1RVxDBKwSwx0a8exRCGgp7wmD4EQit6PwghY%2BIuOb/IShqJlJOkD1YNxHDqUcE5zeZZHhw4kPayoFQ0c%2BXHCdJynpouGV49Z1wOd8pTT5lFW4sVi7RC5l4UucJojRBVBcLml8m4tzltlSwHcgH50vl4cOrd0Fdz5hzJI9hJBAA), smallvec is [495 instructions](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:23,endLineNumber:1,positionColumn:23,positionLineNumber:1,selectionStartColumn:23,selectionStartLineNumber:1,startColumn:23,startLineNumber:1),source:'extern+crate+smallvec%3B%0Ause+smallvec::SmallVec%3B%0A%0Apub+fn+fn2(a:+Option%3Cu32%3E,+b:+Option%3Cu32%3E,+c:+Option%3Cu32%3E)+%7B%09%0A++++++++let+mut+bindgroup_indices:+SmallVec%3C%5Bu32%3B+3%5D%3E+%3D+SmallVec::with_capacity(3)%3B%0A++++++++if+let+Some(mesh_index)+%3D+a++%7B%0A++++++++++++bindgroup_indices.push(mesh_index)%3B%0A++++++++%7D%0A++++++++if+let+Some(skin_index)+%3D+b+%7B%0A++++++++++++bindgroup_indices.push(skin_index)%3B%0A++++++++%7D%0A++++++++if+let+Some(morph_index)+%3D+c+%7B%0A++++++++++++bindgroup_indices.push(morph_index)%3B%0A++++++++%7D%0A++++++++println!!(%22%7B:%3F%7D%22,+%26bindgroup_indices)%3B%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1590,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!((name:smallvec,ver:'180')),options:'-C+opt-level%3D3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.59.0+(Editor+%231)',t:'0')),k:33.830845771144276,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'rustc+1.70.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.59.0+(Compiler+%231)',t:'0')),k:32.83582089552238,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) (although probably some of that is not used, it's hard to tell, but it's clearly worse).

a better alternative: 
```suggestion
        let mut bindgroup_indices: [u32; 3] = Default::default();
        let mut index_count = 0;
        if let Some(mesh_index) = batch_indices.dynamic_offset {
            bindgroup_indices[index_count] = mesh_index;
            index_count += 1;
        }
        if let Some(skin_index) = skin_index {
            bindgroup_indices[index_count] = skin_index.index;
            index_count += 1;
        }
        if let Some(morph_index) = morph_index {
            bindgroup_indices[index_count] = morph_index.index;
            index_count += 1;
        }
        pass.set_bind_group(I, bind_group, &bindgroup_indices[0..index_count]);
```
this is [135 instructions](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATKUcAZPAZMADl3ACNMYgkAdlIAB1QFQjsGFzcPb3jE5IEAoNCWCKiuWKtMGxShAiZiAjT3Tx8yioEqmoI8kPDImMtq2vqMpv6OwK7CnpKASktUE2Jkdg44kzCAaioGDYYAZggmEDWAeTjbAQBSHacTHa9L7FI1sMOTs4ZL69v7x%2BQX05SPjc7jtsFM1udogAhc4aACCMIAnPQCGsWCYUWFAuhgMQ5nEAPpYvCLBSHc4AVkhQMukJ2FIAIuCdoyKZCNKR2RoGTSYfCNEjMCi0SisZhVPi0CZBEzGVydtC4WslcqlXgqGtkWshKg2BA2AoEISGFhVGDLoymEqIQrYSq7crMcacXijfgSazReLJYIGTLUZgDa6xTy4YiVZ6JXNpecvNDmWsuCHbfbwdF6byU2s1RrBVqdZgIAoANaBIOmv3ra0ZzMOrHOkwEonuykR70EX3mtbF0uepNh5WtqMomNxxmJ%2BXV%2B0Q9OKlPZzXa3UsEhxQ2es3x5Cpm012tO3EN13EgMe41iyNS9vklnx5fEVdlvv88Nnr1D8Gxzvjncp6eTu1xMQgQELQDBgJAMZ3FCICXAAYn%2BXg%2BB%2BABsjrYgejbGseCishoAB0eGDpeDJTH2aYcDMtCcOSvCeBwWikKgnAAEpmCiChzAsmAfjsPCkAQmgUTMRYgDsOx4WJklSdJyH6Jwki0YJjGcLwpLsgJ9EUaQcCwEgaAsHEdCROQlD6YZ9BRMQXDkgi7JYAAbseABqeCYAA7icjCcHxNC0AQkSkhAzyaaQjo1AAnt5vBhcQ4VHGE2jlBpfH6WwghHAwtCRSFWBhCYwBOGItCktwvBYCwhjAOIOV4MQSV4PZAZKWK5ToksfHAZgVEhbQeBhMQEUuFgSkEEBLBRTMVAGMACgue5nl0T5sgiGI7BSDIgiKCo6ghboXD6JVICmOY%2Bh9aSsDMGwIAKBVtC0I1yCkI1UTcAAHFoUwzKg/wCCVAC0ThrN9BB/fQjW0OaOy8Kgz1AVg50QDMzRvA4xqDJ4%2B3%2BGMBRFHoCRJG86N49kbydDjkyWF19WtCMRP7cjlQjGT3TFH07R02ztTMxMxRI5xiwSJR1GKSFTEcGsx0EFuXB4TZeEaGsEC4IQJA8VwUy8BpH3CaJ4mvTsGivZI5LRNEXBSJIXjRLJ3UKaQdEMWLqkgOpgkzDpiAgHMBArBYFAQGZRnEMErBLJL0t4dE%2BEMZg%2BBEHDej8IIK3iOtSfyEoahKXtpBuQNcQTXJHA0fbSli0c6K%2B0D6rhwmssIvLisuAZQdqxr/Fuzrtx4fr5KvV4hsIuS5KITsr1F3bDvQyplgux3mmfUXXgi47M9a0JT0BSkICSEAA).

Reply:
Done.